aten/src/ATen/NamedTensorUtils.cpp

```diff
@@ -51,6 +51,95 @@ int64_t dimname_to_position(const Tensor& tensor, Dimname dim) {
   return std::distance(names.begin(), it);
 }
 
+static void report_positional_error(
+    const Dimname& name,
+    const Dimname& other_name,
+    DimnameList names,
+    DimnameList other_names) {
+  // XXX: Can improve message by checking if names are alignable and suggesting workarounds
+  TORCH_CHECK(false,
+      "Names ", name, " and ", other_name, " do not match positionally ",
+      "from the right in names ", names, " and ", other_names, ".");
+}
+
+static void check_for_misalignment(
+    const Dimname& name,
+    DimnameList names,
+    DimnameList other_names) {
+  if (name.is_wildcard()) {
+    return;
+  }
+  auto it = std::find_if(other_names.begin(), other_names.end(),
+      [&](const Dimname& candidate) { return name.can_refer_to(candidate); });
+  // XXX: Can improve message by checking if names are alignable and suggesting workarounds
+  TORCH_CHECK(it == other_names.end(),
+      "Names ", names, " and ", other_names, " are misaligned: name ", name,
+      " appears in a different position from the right.");
+}
+
+// Assumption: A DimnameList can have no duplicate full names with
+// the exception of wildcards
+static std::vector<Dimname> unify_from_right(DimnameList names, DimnameList other_names) {
+  const auto wildcard = Dimname::wildcard();
+  const auto size = std::max(names.size(), other_names.size());
+  auto result = std::vector<Dimname>(size, wildcard);
+
+  auto names_it = names.rbegin();
+  auto other_it = other_names.rbegin();
+  auto result_it = result.rbegin();
+  while (names_it != names.rend() || other_it != other_names.rend()) {
+    // XXX: Don't support tagged names for now. They're a little weird.
+    if (names_it->is_tagged() || other_it->is_tagged()) {
+      TORCH_INTERNAL_ASSERT("unify_from_right: NYI: tagged names.");
+    }
+
+    const auto& name = names_it == names.rend() ? wildcard : *names_it;
+    const auto& other_name = other_it == other_names.rend() ? wildcard : *other_it;
+
+    // Step 1: Check that the names match
+    const auto maybeName = unify(name, other_name);
+    if (!maybeName) {
+      report_positional_error(name, other_name, names, other_names);
+    }
+    *result_it = *maybeName;
+
+    // Step 2: Check that the names are not misaligned
+    if (!names_it->is_normal() || !other_it->is_normal()) {
+      // Let: N = max(len(names), len(other_names))
+      //      K = # of special names among names and other_names.
+      // This search (including the outer loop) is O(N*K) but typically # of dims is small.
+      check_for_misalignment(name, names, other_names);
+      check_for_misalignment(other_name, other_names, names);
+    }
+
+    if (names_it != names.rend()) {
+      ++names_it;
+    }
+    if (other_it != other_names.rend()) {
+      ++other_it;
+    }
+    ++result_it;
+  }
+  return result;
+}
+
+// Assumption: A DimnameList can have no duplicate full names with
+// the exception of wildcards
```

Comment:
I think it'd be helpful to have comments explaining what this function is for, maybe explaining the behavior too or linking to any docs on it

Reply:
Sure, I'll toss some examples in.

Reply:
Added docs in NamedTensorUtils.h